### PR TITLE
Iotbzh/for upstream/add pwm

### PR DESCRIPTION
**Pull request by pmarzin**
Test a PR into iotbzh/main to prepare integration of R-Car PWM driver.

Tested on H3ULCB with pwm0 and pwm4 with blinky_pwm demo. See bellow picture (for fun).
PWM_blinky_test2

### CODE REVIEW

File: drivers/pwm/pwm_rcar.c [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+#define LOG_LEVEL CONFIG_PWM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pwm_rcar);
+
+/* Helper macros for PWM */
+#define DEV_PWM_CFG(dev)        ((const struct pwm_rcar_cfg *)(dev)->config)
+#define DEV_PWM_DATA(dev)       ((struct pwm_rcar_data *)(dev)->data)
+
+/* PWM Controller capabilities */
+#define RCAR_PWM_MAX_CYCLE      1023U
```
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:27]: can you align all the constant definition ?

File: drivers/pwm/pwm_rcar.c [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+#define RCAR_PWM_CR_CC_SHIFT            16
+#define RCAR_PWM_CR_CCMD                BIT(15)         /* Clock Control Frequency Division Mode */
+#define RCAR_PWM_CR_SYNC                BIT(11)
+#define RCAR_PWM_CR_SS                  BIT(4)          /* Single Pulse Output */
+#define RCAR_PWM_CR_EN                  BIT(0)          /* Channel Enable */
+
+/* PWMCNT (PWM Count Register) */
+#define RCAR_PWM_CNT_CYC_MASK           0x03ff0000      /* PWM Cycle */
+#define RCAR_PWM_CNT_CYC_SHIFT          16
+#define RCAR_PWM_CNT_PH_MASK            0x000003ff      /* PWM Hight-Level Period */
```
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c]:
```suggestion
#define RCAR_PWM_CNT_PH_MASK            0x000003ff      /* PWM High-Level Period */
```

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml [hunk 3]
```diff
@@ -11,6 +11,7 @@ supported:
   - gpio
   - clock_control
   - uart
+  - pwm
```
Comment (julien-massot) [inline boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml]: does unit-test and twister works ?

Comment (julien-massot) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml]: and this board modification should goes in a separate commit

File: drivers/pwm/pwm_rcar.c (hunk 2 above)
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c:45]: realignement has been set by `uncrustify` tool... :/

File: drivers/pwm/pwm_rcar.c (hunk 1 above)
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c:27]: realignement has been set by uncrustify tool... :/

File: drivers/pwm/pwm_rcar.c (hunk 2 above)
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: Done. thanks

File: drivers/pwm/pwm_rcar.c (hunk 1 above)
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c:27]: Uncrustify matted ! haha ;) thanks !

File: drivers/pwm/pwm_rcar.c [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+ */
+
+#define DT_DRV_COMPAT renesas_rcar_pwm
+
+#include <errno.h>
+#include <device.h>
+#include <devicetree.h>
+#include <soc.h>
+#include <drivers/pwm.h>
+#include <drivers/clock_control.h>
```
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:14]: include order sounds odd either alphabetic order or reversed christmas tree

**Pull request closed without merging**